### ISSUE TITLE
WSL2 + Ubuntu 22.04.1:  xdg-open: no method available for opening "https://www...."

### ISSUE THREAD

**Issue by ramonduraes** [microsoft/WSL#8892]
### Version

Microsoft Windows [´veersion 10.0.19042.1706]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.10.102.1

### Distro Version

Ubuntu  22.04.1

### Other Software

Try
xdg-open https://microsoft.com

result
/usr/bin/xdg-open: 882: www-browser: not found
/usr/bin/xdg-open: 882: links2: not found
/usr/bin/xdg-open: 882: elinks: not found
/usr/bin/xdg-open: 882: links: not found
/usr/bin/xdg-open: 882: lynx: not found
/usr/bin/xdg-open: 882: w3m: not found
xdg-open: no method available for opening 'https://microsoft.com'

### Repro Steps

Try
xdg-open https://microsoft.com

### Expected Behavior

Open windows browser

### Actual Behavior

/usr/bin/xdg-open: 882: www-browser: not found
/usr/bin/xdg-open: 882: links2: not found
/usr/bin/xdg-open: 882: elinks: not found
/usr/bin/xdg-open: 882: links: not found
/usr/bin/xdg-open: 882: lynx: not found
/usr/bin/xdg-open: 882: w3m: not found
xdg-open: no method available for opening 'https://microsoft.com'

### Diagnostic Logs

/usr/bin/xdg-open: 882: www-browser: not found
/usr/bin/xdg-open: 882: links2: not found
/usr/bin/xdg-open: 882: elinks: not found
/usr/bin/xdg-open: 882: links: not found
/usr/bin/xdg-open: 882: lynx: not found
/usr/bin/xdg-open: 882: w3m: not found
xdg-open: no method available for opening 'https://microsoft.com'

**Comment by benhillis** [comment]
I think you may have to configure xdg-open to do this, we don't do this by default.

**Comment by sabbour** [comment]
`az login` was working for me before, but now I'm facing this issue.

**Comment by fghamsary** [comment]
I have the same issue, just installed a fresh Ubuntu 22.04 with service and GUI enabled, and the interop with browser is not working anymore on this version, there is something wired going on when using systemd with new WSL and GUI

**Comment by fghamsary** [comment]
Came accros this one: https://github.com/microsoft/WSL/issues/8795
It seems like on eariler versions there was a package installed by default wslu which is not available by default on 22.04.
It had a wslview shell command which did the work, so by installing it, you can have the functionality back, but not sure of the drawbacks.

**Comment by AdamGlass** [comment]
wslu works.  This really feels like a problem WSL should own -- installing the component, ensuring that the user knows the component could be installed, or forcing it to be installed due to some dependency.

it's not even mentioned here: https://learn.microsoft.com/en-us/windows/wsl/install

**Comment by lgmorand** [comment]
Same issue here. the WSL 22.04 was working fine. I did a apt-upgrade and then, the console is now broken when trying to do "az login")

**Comment by belainehi** [comment]
I had a similar issue, considering if installing one of the "not found" packages would help. Haven't found a solution yet.

**Comment by Biswa96** [comment]
Try to add an web browser path with `BROWSER` environment variable before launching xdg-open. For example, let assume Firefox is installed in Windows as `C:\Firefox\firefox.exe`. Then set the environment variable as `export BROWSER=/mnt/c/Firefox/firefox.exe` and run xdg-open as usual.

wslpath can also help to covert the path style.

Source: https://cgit.freedesktop.org/xdg/xdg-utils/tree/scripts/xdg-open.in

**Comment by AndreasNasman** [comment]
@Biswa96 Thanks for the workaround! 👍

I had to create a symlink with `ln -s /mnt/c/path/to/browser.exe` and then set `export BROWSER=~/path/to/symlink` as my browser is installed in a directory with spaces (`Program Files (86)`), which `xdg-open` can't handle.
https://unix.stackexchange.com/a/715404/571729

**Comment by LucianoCrocco** [comment]
> @Biswa96 Thanks for the workaround! 👍
> 
> I had to create a symlink with `ln -s /mnt/c/path/to/browser.exe` and then set `export BROWSER=~/path/to/symlink` as my browser is installed in a directory with spaces (`Program Files (86)`), which `xdg-open` can't handle. https://unix.stackexchange.com/a/715404/571729

Thanks both of you, it worked having 401 error.

**Comment by jiasli** [comment]
I am the developer of Azure CLI authentication.

Because of this change, Python's built-in module [`webbrowser`](https://docs.python.org/3/library/webbrowser.html) stops working on Ubuntu 22.04 in WSL:

```
$ python -c "import webbrowser; print(webbrowser.open('https://login.microsoftonline.com/'))"
True
(py310) user2@DESKTOP-A79F1:~$ /usr/bin/xdg-open: 882: x-www-browser: not found
/usr/bin/xdg-open: 882: firefox: not found
/usr/bin/xdg-open: 882: iceweasel: not found
/usr/bin/xdg-open: 882: seamonkey: not found
/usr/bin/xdg-open: 882: mozilla: not found
/usr/bin/xdg-open: 882: epiphany: not found
/usr/bin/xdg-open: 882: konqueror: not found
/usr/bin/xdg-open: 882: chromium: not found
/usr/bin/xdg-open: 882: chromium-browser: not found
/usr/bin/xdg-open: 882: google-chrome: not found
/usr/bin/xdg-open: 882: www-browser: not found
/usr/bin/xdg-open: 882: links2: not found
/usr/bin/xdg-open: 882: elinks: not found
/usr/bin/xdg-open: 882: links: not found
/usr/bin/xdg-open: 882: lynx: not found
/usr/bin/xdg-open: 882: w3m: not found
xdg-open: no method available for opening 'https://login.microsoftonline.com/'
```

This affects MSAL's `PublicClientApplication.acquire_token_interactive`: https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/64141cacdb4ea3a02a2b3cb45bc8fb3be4f93f36/msal/oauth2cli/authcode.py#L58

Then affects Azure CLI's `az login`: https://github.com/Azure/azure-cli/blob/f0c87ca820e0bd3f9dd403ef96c21ed49b734d16/src/azure-cli-core/azure/cli/core/auth/identity.py#L153

**Comment by lgmorand** [comment]
@jiasli what does that mean ? that we should consider az cli not working properly anymore on WSL 22.04 ?

**Comment by lgmorand** [comment]
@kierzorz thanks but that's not the point. WSL is configured and install by MS, az CLI is packaged by MS. Can't we expect either a fix upstream or a built-in workaround. For instance, WSL is installed on Windows where edge is here by default (even if not default browser), so for me, the env could be setup to point to edge install path. it would be sufficient to open a web page for auth

**Comment by leslie-corbalt** [comment]
I just installed Ubuntu 22.04. I had to do 2 things:
1. install xdg utilities (`sudo apt install xdg-utils`) and 
2. install wslu following the instructions here: https://wslutiliti.es/wslu/install.html (Scroll down to `**Ubuntu Warning**`.  Follow the steps for PPA version:
`sudo add-apt-repository ppa:wslutilities/wslu`
`sudo apt update`
`sudo apt install wslu`

**Comment by cianetech** [comment]
@leslie-corbalt I had a lot of issues because of `xdg-open` not really working. Thanks to you  and to the other people who mentioned about WSLU. It worked for me. I followed the steps you mentioned above.

**Comment by ackalker** [comment]
I'm using this on WSL2 Ubuntu with default Windows integrations (such as ability to run .exe from Linux) enabled:

```shell
$ sudo apt install xdg-utils
$ export BROWSER="cmd.exe /c start"
$ xdg-open https://www.microsoft.com/
'\\wsl.localhost\Ubuntu\home\<username_redacted>'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
```

This opens https://www.microsoft.com/ in my default Windows browser. It also appears to work with URL with query strings (`?foo`) and fragments (`#bar`) although I haven't tested it extensively, there could be problems with quoting across platforms. The 'UNC paths are not spported' appears to be harmless.

**Comment by leandiazz** [comment]
> Same issue here. the WSL 22.04 was working fine this morning. I did a apt-upgrade and then, the console is now broken when trying to do "az login"
> 
> I did "sudo apt install wslu" like mentioned and it seems to fix it. pretty weird what broke it

This work for me. Thanks!

**Comment by tgiacaman** [comment]
> I just installed Ubuntu 22.04. I had to do 2 things:
> 
> 1. install xdg utilities (`sudo apt install xdg-utils`) and
> 2. install wslu following the instructions here: https://wslutiliti.es/wslu/install.html (Scroll down to `**Ubuntu Warning**`.  Follow the steps for PPA version:
>    `sudo add-apt-repository ppa:wslutilities/wslu`
>    `sudo apt update`
>    `sudo apt install wslu`

I'm using WSL2 on Windows using Ubuntu v22.04.5 LTS. -->  I just execute $ sudo apt install xdg-utils and the call to a browser was solved and the "gio warning" dissapears

**Comment by copdips** [comment]
> [@Biswa96](https://github.com/Biswa96) Thanks for the workaround! 👍
> 
> I had to create a symlink with `ln -s /mnt/c/path/to/browser.exe` and then set `export BROWSER=~/path/to/symlink` as my browser is installed in a directory with spaces (`Program Files (86)`), which `xdg-open` can't handle. https://unix.stackexchange.com/a/715404/571729

you can do it like this, with quotes around the part having spaces, but not the full path
```bash
export BROWSER=/mnt/c/'Program Files (x86)'/Microsoft/Edge/Application/msedg
e.exe
```

**Comment by khacnha** [comment]
For those who are lazy, just copy the displayed link in command line and open it in the browser, the command can still receive feedback after you finish working in the browser.

**Comment by vadimkantorov** [comment]
same problem in WSL1 + Ubuntu 24.04. It would be nice if somehow WSL provided some help in this

**Comment by slatian** [comment]
xdg-utils maintianer here:

Thanks to @vadimkantorov for [opening an upstream bug](https://gitlab.freedesktop.org/xdg/xdg-utils/-/issues/277)

TL;DR The xdg-utils aren't the only program that makes use of the `BROWSER` variable, so xdg-open tries to stay compatible resulting in what looks like ignoring the problem.

The official xdg-utils workaround is creating a shell script that opens your favorite browser and put that into the `BROWSER` variable. (Will be in the manpage on the next release too)

Example shell script:
```sh
#!/bin/sh

exec "/mnt/c/Program Files (x86)/my/favorite/browser.exe" "$@"
```

In case anyone from the WSL project  is reading this: I'd be great to have someone I can ping to get WSL knowledge and who can test changes to avoid  breakages like this in the future.

**Comment by trallnag** [comment]
Note that https://github.com/wslutilities/wslu has been discontinued.

And note that https://launchpad.net/ubuntu/+source/wslu is outdated and the latest version of wslu is only available in the aforementioned GitHub repo or as PPA https://launchpad.net/~wslutilities/+archive/ubuntu/wslu

**Comment by gabriellaXL** [comment]
It worked after I downloaded wslu.
sudo add-apt-repository ppa:wslutilities/wslu
sudo apt update
sudo apt install wslu

**Comment by dellch** [comment]
I can at least confirm `xdg-open` (after installing it with `sudo apt install xdg-utils`) works out of the box in WSL2 with Ubuntu-26.06 without installing any additional packages like `wslu` .
